### PR TITLE
Python client changes for json inputs

### PR DESCRIPTION
**Pull request by seanshi-scale**
Python client handles json/serialized inputs + outputs, also add a `bundle_location_fn` notion.

Added an `EndpointRequest` class to encapsulate the options for a request to an endpoint

### CODE REVIEW

File: nucleus/deploy/model_endpoint.py [hunk 1]
```diff
@@ -57,28 +90,27 @@ def predict_batch(
...
                 endpoint_id=self.endpoint_id,
-                url=inner_url,
+                url=request.url,
+                args=request.args,
+                return_pickled=request.return_pickled,
             )
-            return inner_url, inner_inference_request
+            if request.request_id is not None:
+                request_key = request.request_id
+            elif request.url is not None:
```
Comment (seanshi-scale) [inline nucleus/deploy/model_endpoint.py:104]: maybe this stuff really isn't necessary :P

File: nucleus/deploy/model_endpoint.py [hunk 2]
```diff
@@ -1,12 +1,41 @@
...
 from collections import Counter
-from typing import Dict, Optional, Sequence
+from typing import Any, Dict, Optional, Sequence
 
 TASK_PENDING_STATE = "PENDING"
 TASK_SUCCESS_STATE = "SUCCESS"
 TASK_FAILURE_STATE = "FAILURE"
 
 
+class EndpointRequest:
```
Comment (seanshi-scale) [inline nucleus/deploy/model_endpoint.py:10]: since we have a function to predict on multiple inputs, I added this so people could pass in a mix of urls/jsons as inputs. Not sure if that'll happen

**Review by ivan-khomyakov-sc — APPROVED** [review]

**Review by yixu34 — COMMENTED** [review]
Don't merge yet still commenting

File: nucleus/deploy/model_endpoint.py (hunk 2 above)
Comment (yixu34) [reply nucleus/deploy/model_endpoint.py:10]: Might be good to add an `EndpointResponse` class for symmetry?

File: nucleus/deploy/model_endpoint.py [hunk 3]
```diff
@@ -15,8 +44,13 @@ def __init__(self, endpoint_id: str, client):
     def __str__(self):
         return f"SyncModelEndpoint <endpoint_id:{self.endpoint_id}>"
 
-    def predict(self, url):
-        return self.client.sync_request(self.endpoint_id, url)
+    def predict(self, request: EndpointRequest):
```
Comment (yixu34) [inline nucleus/deploy/model_endpoint.py:47]: Per above, `-> EndpointResponse` might be a handy result and annotation.

File: nucleus/deploy/model_endpoint.py (hunk 1 above)
Comment (yixu34) [reply nucleus/deploy/model_endpoint.py:104]: wdym?

File: nucleus/deploy/model_endpoint.py [hunk 4]
```diff
@@ -1,12 +1,41 @@
...
+class EndpointRequest:
+    """
+    Represents a single request to either a SyncModelEndpoint or AsyncModelEndpoint.
+    Parameters:
+        url: A url to some file that can be read in to a ModelBundle's predict function. Can be an image, raw text, etc.
+        args: A Dictionary with arguments to a ModelBundle's predict function. If the predict function has signature
+            predict_fn(foo, bar), then the keys in the dictionary should be 'foo' and 'bar'. Values must be native Python
+            objects.
+        return_pickled: Whether the output should be a pickled python object, or directly returned serialized json
+        request_id: A user-specifiable id for requests. Should be unique.
```
Comment (yixu34) [inline nucleus/deploy/model_endpoint.py:19]: When/why would a user specify this? I was imagining this would be populated by us server-side.

File: nucleus/deploy/model_endpoint.py [hunk 5]
```diff
@@ -1,12 +1,41 @@
...
+    """
+
+    def __init__(
+        self,
+        url: Optional[str] = None,
+        args: Optional[Dict] = None,
```
Comment (yixu34) [inline nucleus/deploy/model_endpoint.py:25]: Wonder if we should have these two things be a `Union` (we'd still have to do the XOR check because `Union` is really just a hint, from a type annotation perspective).

File: nucleus/deploy/model_endpoint.py [hunk 6]
```diff
@@ -1,12 +1,41 @@
...
+        return_pickled: Optional[bool] = True,
+        request_id: Optional[str] = None,
+    ):
+        if url is None and args is None:
```
Comment (yixu34) [inline nucleus/deploy/model_endpoint.py:29]: Related to the `Union` comment, think this is happening enough where we'd want a helper function.

File: nucleus/deploy/model_endpoint.py (hunk 4 above)
Comment (seanshi-scale) [reply nucleus/deploy/model_endpoint.py:19]: The thing this `request_id` is trying to solve is that a user needs to be able to associate model inference results with a given model input. For example, for the nucleus integration we need to maintain some map from s3URL of input images to task results, so this request_id can be set to that s3URL. Before, I was implicitly using the url as the request_id, but we don't necessarily have urls anymore.

I suppose that we could just keep the association implicitly in the order of some list somewhere but then the user would have to do that too.

Comment (yixu34) [reply nucleus/deploy/model_endpoint.py:19]: Ok it sounds like you're basically using this as a `reference_id` then? Can we just have our client generate a UUID? Then the nucleus integration (which one day will be server-side) can just maintain that `Request` -> `Response` mapping, with the hash key being this UUID.

File: nucleus/deploy/model_endpoint.py (hunk 1 above)
Comment (seanshi-scale) [reply nucleus/deploy/model_endpoint.py:104]: there's a lot of logic to try to figure out a reasonable request_id if the user didn't pass one in, e.g. use the url, or the args, which seems "a bit extra"

File: nucleus/deploy/model_endpoint.py (hunk 5 above)
Comment (seanshi-scale) [reply nucleus/deploy/model_endpoint.py:25]: I sorta am leaning away from putting the two into a `Union`. We'd have to keep track of whether the data represents a url or args (which we technically could do by looking at their types, but this feels wrong), and keeping them separate seems like the easiest way to do this. The stuff stored at url/args gets put into separate fields in the eventual http request's json anyway, so I think it's simpler to keep things separate.

File: nucleus/deploy/model_endpoint.py (hunk 1 above)
Comment (yixu34) [reply nucleus/deploy/model_endpoint.py:104]: Oh ok, as per my other comment ([18 hours ago](https://github.com/scaleapi/nucleus-python-client/pull/232#discussion_r818259959)), can we just generate a UUID?

File: nucleus/deploy/model_endpoint.py (hunk 4 above)
Comment (seanshi-scale) [reply nucleus/deploy/model_endpoint.py:22]: I left in the ability for clients to specify their own `request_id`s, but otherwise the client will generate a uuid for the `request_id`

Comment (yixu34) [reply nucleus/deploy/model_endpoint.py:22]: I would err on the more restrictive side and not allow `request_id`s, because I'm imagining a hapless user passing in "42" hardcoded and never changing it.

File: nucleus/deploy/model_endpoint.py (hunk 5 above)
Comment (yixu34) [reply nucleus/deploy/model_endpoint.py:29]: Yup that makes sense.

File: nucleus/deploy/nucleus_integration.py [hunk 7]
```diff
@@ -26,7 +26,7 @@ class NucleusDatasetInferenceRun:
     # For the demo, we will need our Nucleus Dataset to have `image_location`s in s3://scale-ml-hosted-model-inference
     def __init__(
         self,
-        hmi_async_job: AsyncModelEndpointResponse,
+        hmi_async_job: AsyncModelEndpointBatchResponse,
```
Comment (yixu34) [inline nucleus/deploy/nucleus_integration.py:29]: Not really part of this PR but maybe we can take a pass at removing all "hmi" stuff from user-facing names?

File: nucleus/deploy/model_endpoint.py (hunk 4 above)
Comment (seanshi-scale) [reply nucleus/deploy/model_endpoint.py:22]: (probably should have specified this but) The uniqueness condition is actually more permissive than "globally unique", but instead is "the EndpointRequests producing a given BatchResponse must have distinct `request_ids`". I think that mostly gets rid of the hardcoding problem? (i.e. it doesn't matter if the user runs the script twice with the same value of `request_id`, but if the user hardcodes a value for `request_id` somewhere and copies the code that would be a problem)

I do think we should allow users to set `request_id`s though, since otherwise it's much more annoying for users to keep track of an association from `EndpointRequest`/`EndpointResponse` to something a user cares about, e.g. if the user wants to run inference on N images, they need to keep track of which `EndpointResponse` gets associated with which image. 

With user-specified `request_id`s the user can do `endpoint_response_for_image_i = job.responses[request_id_for_image_i]`; with non-user-specified `request_ids` the user would need to do 
```
# Go through the list of EndpointRequests to find the request_id corresponding to a given image, comparing either `url` or `args`. Comparing values seems reasonable in the `url` case but ugly in the `args` case.
# job.responses[the request_id we found]
```
or use the ordering of `EndpointRequest` fed in and the corresponding ordering of `EndpointResponses` (currently `EndpointResponses` are returned in a dictionary though)
(or something I haven't thought of)

One simple thing we could do to catch errors earlier is just to check that the user passes a set of distinct `request_id`s to the batch_predict function at the very least

Comment (yixu34) [reply nucleus/deploy/model_endpoint.py]: Oh I see. I think this is happening because the client library is still operating under the paradigm where batch jobs are tied to endpoints instead of being standalone. In the eventual state, I would expect the batch request to be its own class (not a `predict_batch` method off of an endpoint class), and result would be a url to a file/directory that contains outputs, with the relative order not being guaranteed anyway. In other words, this is needed only because have to support the current janky client-side integration. In that case I guess this is fine, but prior to launch I definitely would want `request_id` to be gone altogether from the client, so please add a TODO.

I was tempted to suggest that we call this `reference_id`, but that's misleading in a _different_ way (Nucleus actually persists those, so for us to use them as ephemeral client-side IDs could be confusing due to inconsistency).

Comment (seanshi-scale) [reply nucleus/deploy/model_endpoint.py]: Yeah it's pretty much there because of the async batch class

**Review by yixu34 — APPROVED** [review]
